web/pandas/pdeps/0017-backwards-compatibility-and-deprecation-policy.md

```diff
@@ -0,0 +1,75 @@
+# PDEP-17: Backwards compatibility and deprecation policy
+
+- Created: 27 June 2024
+- Status: Under discussion
+- Discussion: [#59125](https://github.com/pandas-dev/pandas/issues/59125)
+- Author: [Abdulaziz Aloqeely](https://github.com/Aloqeely)
+- Revision: 1
+
+## Abstract
+
+This PDEP defines pandas' backwards compatibility and deprecation policy.
+
+## Motivation
+
+Having a clear backwards compatibility and deprecation policy is crucial to having a healthy ecosystem. We want to ensure users can rely on pandas being stable while still allowing the library to evolve.
+
+This policy will ensure that users have enough time to deal with deprecations while also minimizing disruptions on downstream packages' users.
+
+## Scope
+
+This PDEP covers pandas' approach to backwards compatibility and the deprecation and removal process.
+
+## Background
+
+pandas uses a loose variant of semantic versioning.  
+A pandas release number is written in the format of ``MAJOR.MINOR.PATCH``.
+
+## General policy
+
+This policy applies to the [public API][1]. Anything not part of the [public API][1] or is marked as "Experimental" may be changed or removed at anytime.
+
+- Breaking backwards compatibility should benefit more than it harms users.
+- Breaking changes should go through a deprecation cycle before being implemented.
+- Breaking changes should only occur in major releases.
+- No deprecations should be introduced in patch releases.
+- Deprecated functionality should remain unchanged for at least 6 months before being changed or removed.
+
+Some bug fixes may require breaking backwards compatibility. In these cases, a deprecation cycle is not necessary. However, bug fixes which have a large impact on users might be treated as a breaking change. Whether or not a change is a bug fix or an API breaking change is a judgement call.
+
+## Deprecation process
+
+Deprecation provides a way to warn developers and give them time to adapt their code to the new functionality before the old behavior is eventually removed.
+
+A deprecation's warning message should:
+- Provide information on what is changing.
+- Mention how to achieve similar behavior if an alternative is available.
+- Include the version in which the deprecation will be enforced.
```

Comment:
Currently we don't do this (at least, often I think). It seems to me including the version is often not particularly helpful for a user, they should change their code to avoid the deprecation as soon as is convenient and that is independent of the version it will be enforced. And the version we provide here could always end up being incorrect.

Reply:
I took this from the [Version policy](https://pandas.pydata.org/docs/development/policies.html) page but I agree there is not much value in doing this.

Reply:
@pandas-dev/pandas-core - any opposition to removing the version where the deprecation will be enforced? I wasn't aware of this in the version policy, it seems most often we say something like "will be removed in a future version".

Reply:
I'm guessing no comments means no opposition, so I'll remove this.